src/mem/alloc.rs

```diff
@@ -0,0 +1,1063 @@
+    }
+
+    fn encode_order(&self, order: usize) -> PageStorageType {
+        PageStorageType(self.0 | ((order as u64) & PAGE_ORDER_MASK) << PAGE_TYPE_SHIFT)
+    }
+
+    fn encode_next(&self, next_page: usize) -> PageStorageType {
+        PageStorageType(self.0 | (next_page as u64) << PAGE_FREE_NEXT_SHIFT)
+    }
+
+    fn encode_slab(slab: VirtAddr) -> Self {
+        PageStorageType(PAGE_TYPE_SLABPAGE | slab.as_u64() & PAGE_TYPE_SLABPAGE_MASK)
+    }
+}
+
+const PAGE_TYPE_SHIFT: u64 = 4;
+const PAGE_TYPE_MASK: u64 = (1u64 << PAGE_TYPE_SHIFT) - 1;
+
+const PAGE_TYPE_FREE: u64 = 0;
+const PAGE_FREE_NEXT_SHIFT: u64 = 12;
+const PAGE_FREE_NEXT_MASK: u64 = !((1u64 << PAGE_FREE_NEXT_SHIFT) - 1);
+
+const PAGE_TYPE_ALLOCATED: u64 = 1;
+
+const PAGE_ORDER_MASK: u64 = (1u64 << (PAGE_FREE_NEXT_SHIFT - PAGE_TYPE_SHIFT)) - 1;
+
+// SLAB pages are always order-0
+const PAGE_TYPE_SLABPAGE: u64 = 2;
+const PAGE_TYPE_SLABPAGE_MASK: u64 = !PAGE_TYPE_MASK;
+
+const PAGE_TYPE_COMPOUND: u64 = 3;
+
+const PAGE_TYPE_RESERVED: u64 = (1u64 << PAGE_TYPE_SHIFT) - 1;
+
+struct FreeInfo {
+    next_page: usize,
+    order: usize,
+}
+
+impl FreeInfo {
+    pub fn encode(&self) -> PageStorageType {
+        PageStorageType::new(PAGE_TYPE_FREE)
+            .encode_order(self.order)
+            .encode_next(self.next_page)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        let next = ((mem.0 & PAGE_FREE_NEXT_MASK) >> PAGE_FREE_NEXT_SHIFT) as usize;
+        let order = ((mem.0 >> PAGE_TYPE_SHIFT) & PAGE_ORDER_MASK) as usize;
+        FreeInfo {
+            next_page: next,
+            order: order,
+        }
+    }
+}
+
+struct AllocatedInfo {
+    order: usize,
+}
+
+impl AllocatedInfo {
+    pub fn encode(&self) -> PageStorageType {
+        PageStorageType::new(PAGE_TYPE_ALLOCATED).encode_order(self.order)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        let order = ((mem.0 >> PAGE_TYPE_SHIFT) & PAGE_ORDER_MASK) as usize;
+        AllocatedInfo { order: order }
+    }
+}
+
+struct SlabPageInfo {
+    slab: VirtAddr,
+}
+
+impl SlabPageInfo {
+    pub fn encode(&self) -> PageStorageType {
+        PageStorageType::encode_slab(self.slab)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        SlabPageInfo {
+            slab: VirtAddr::new(mem.0 & PAGE_TYPE_SLABPAGE_MASK),
+        }
+    }
+}
+
+struct CompoundInfo {
+    order: usize,
+}
+
+impl CompoundInfo {
+    pub fn encode(&self) -> PageStorageType {
+        PageStorageType::new(PAGE_TYPE_COMPOUND).encode_order(self.order)
+    }
+
+    pub fn decode(mem: PageStorageType) -> Self {
+        let order = ((mem.0 >> PAGE_TYPE_SHIFT) & PAGE_ORDER_MASK) as usize;
+        CompoundInfo { order: order }
+    }
+}
+
+struct ReservedInfo {}
+
+impl ReservedInfo {
+    fn encode(&self) -> PageStorageType {
+        PageStorageType::new(PAGE_TYPE_RESERVED)
+    }
+
+    pub fn decode(_mem: PageStorageType) -> Self {
+        ReservedInfo {}
+    }
+}
+
+enum Page {
```

Comment:
I don't think Page is a very good name. It may add confusion with Page as defined in other crates, and also it isn't really a page... it's more like page metadata. Maybe call it SvsmPageInfo?